### PR TITLE
Remove unwanted exceptions from @inlinable checking

### PR DESCRIPTION
**Pull request by jrose-apple**
- Treat protocol requirements just like associated types in how they inherit usable-from-inline-ness, rather than special-casing them in the check for inlinable code.

- Enum elements are already treated this way, so we can remove a redundant check for that.

- Finally, start enforcing that `dynamic` declarations need to be `@usableFromInline` to be used in inlinable functions...in Swift 5 mode or resilient code. I didn't even add a warning in Swift 4/4.2 because it was illegal to use `@usableFromInline` on a `dynamic` declaration in Swift 4.2. (Oops.)

See also #19548, which would reasonably allow us to add that warning.

### CODE REVIEW

File: lib/Sema/ResilienceDiagnostics.cpp [hunk 1]
```diff
@@ -111,21 +111,6 @@ bool TypeChecker::diagnoseInlinableDeclRef(SourceLoc loc,
...
-  // Protocol requirements are not versioned because there's no
-  // global entry point.
-  if (isa<ProtocolDecl>(D->getDeclContext()) &&
-      D->isProtocolRequirement())
-    return false;
-
-  // Dynamic declarations are not versioned because there's no
-  // global entry point.
-  if (D->isDynamic())
-    return false;
```
Comment (jrose-apple) [inline lib/Sema/ResilienceDiagnostics.cpp]: @slavapestov I really don't know what to do about this. Fixing this is source-breaking, but 4.2 rejects `@usableFromInline` on `dynamic` declarations.

Comment (slavapestov) [reply lib/Sema/ResilienceDiagnostics.cpp]: Does @usableFromInline matter for dynamic? There’s no public entry point.

Comment (jrose-apple) [reply lib/Sema/ResilienceDiagnostics.cpp]: `-swift-version 4.2` has shipped. We can't do changes there. I guess we'll have to limit it to 5 mode.

We definitely can't print referenced internal dynamic methods; they may be using internal types. Actually, that's just a hole in general without this.

**Comment by jrose-apple** [comment]
Okay, ready for review.

@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
\*sigh\* I always forget the `-enable-objc-interop`.

**Comment by jrose-apple** [comment]
Source compat suite: [Debug](
https://ci.swift.org/job/swift-PR-source-compat-suite-debug/148/), [Release](
https://ci.swift.org/job/swift-PR-source-compat-suite/1919/)

@swift-ci Please test

**Comment by jrose-apple** [comment]
Oops, I included the test case from #19440.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@slavapestov Ping

**Review by slavapestov — APPROVED** [review]

**Comment by jrose-apple** [comment]
Thanks!

Just to be sure…

@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test Linux

**Comment by jrose-apple** [comment]
Hm, actually, I guess the ABI stability checking change is "correct" here, in that we weren't recording these things properly and now we are. @nkcsgexi, @airspeedswift, do you want me to add these to the expected list, or is it better to regenerate the baseline?

**Comment by airspeedswift** [comment]
I'm good with either. Personally, I think we're going to want to regenerate the baseline at intervals for the next short while anyway, so it'll probably only be temporary if you put them in the expected list.

**Comment by nkcsgexi** [comment]
I agree with Ben here. Very likely we will regenerate baseline after other fixes soon, so putting them in the expected list temporarily is fine. When parsable interface is ready, we should checking parsable interface as baseline so we never need to update it.

**Pull request merged**